### PR TITLE
User can make, view, and cancel orders

### PR DESCRIPTION
**Pull request by dtinianow**
User can make reservation
User can view an individual order show page
User can cancel an order

### CODE REVIEW

**Comment by Automatic365** [comment]
This functionality is a game changer

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -3,24 +3,27 @@ class OrdersController < ApplicationController
...
-    @orders = Order.all
+    @orders = current_user.orders
   end
 
   def create
     order = current_user.orders.create
-    @cart.find_items.each do |item|
-      order.order_items.create(item_id: item.id, title: item.title, quantity: @cart.contents[item.id.to_s], price: item.price)
+    @cart.find_items.each do |property|
+      order.reservations.create(property_id: property.id,
```
Comment (pwentz) [inline app/controllers/orders_controller.rb]: Nice use of strptime. Next time troy to add newline before first argument on multi-line argument method calls. That way the args are centered and become easier to read. Overall great work though, nice manipulation with the destroy method.

File: app/models/cart.rb [hunk 2]
```diff
@@ -8,21 +8,25 @@ def initialize(contents)
...
-      occupancy: params[:occupancy] }
+                                            end_date: params[:end_date],
+                                            occupancy: params[:occupancy] }
   end
 
-  def count_of(item_id)
-    contents[item_id.to_s]
-  end
+  # def count_of(item_id)
+  #   contents[item_id.to_s]
```
Comment (Yoyo2Code) [inline app/models/cart.rb]: We need to figure out what to do about those inline comments

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -36,6 +39,6 @@ def order_total(order)
   def destroy
     @order = current_user.orders.find(params[:id])
     @order.update(status: "cancelled")
```
Comment (pwentz) [inline app/controllers/orders_controller.rb]: Keeping them in the db for order history. Nice play.

File: app/models/cart.rb [hunk 4]
```diff
@@ -8,21 +8,25 @@ def initialize(contents)
...
+  # end
 
   def total
     contents.keys.count
   end
 
   def find_items
     contents.map do |id, data|
-      CartItem.new(id, data["occupancy"], data["starting_date"], data["end_date"])
+      CartItem.new(id,
```
Comment (pwentz) [inline app/models/cart.rb]: Here would be a great opportunity to move id to the next line.

File: spec/good_tests/features/visitor_sees_all_properties_in_city_spec0.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+                      description: "Small people",
+                      price_per_guest: 10.00,
+                      image_path: "https://c2.staticflickr.com/4/3597/5790914660_1733d9253f_b.jpg",
+                      max_occupancy: 2,
+                      location: location,
+                      business: business)
+
+    visit '/location/new-york'
+
+    expect(page).to have_content("New York")
```
Comment (Yoyo2Code) [inline spec/good_tests/features/visitor_sees_all_properties_in_city_spec0.rb]: Nice job with all the validations!

File: app/controllers/orders_controller.rb [hunk 6]
```diff
@@ -3,24 +3,27 @@ class OrdersController < ApplicationController
...
 
   def create
     order = current_user.orders.create
-    @cart.find_items.each do |item|
-      order.order_items.create(item_id: item.id, title: item.title, quantity: @cart.contents[item.id.to_s], price: item.price)
+    @cart.find_items.each do |property|
+      order.reservations.create(property_id: property.id,
+                                price: property.subtotal,
+                                number_of_guests: property.occupancy,
+                                starting_date: DateTime.strptime(property.starting_date, "%m/%d/%Y"),
```
Comment (Automatic365) [inline app/controllers/orders_controller.rb]: strptime. Need to remember that

File: spec/good_tests/features/user_can_cancel_order_spec.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    fill_in 'Username', with: user.username
+    fill_in 'Password', with: user.password
+
+    within('.login-form') do
+      click_on "Login"
+    end
+
+    visit property_path(property, business_name: property.business.slug)
+
+    find('#occupancy').find(:xpath, 'option[2]').select_option
```
Comment (pwentz) [inline spec/good_tests/features/user_can_cancel_order_spec.rb]: super clean way to fetch that element 👍

File: spec/good_tests/features/user_can_view_past_orders_spec.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    expect(page).to have_content "08/30/2016"
+    expect(page).to have_content "Check-Out"
+    expect(page).to have_content "09/05/2016"
+    expect(page).to have_content "Total Cost"
+    expect(page).to have_content "$20,000.00"
+    expect(page).to have_content "Status"
+    expect(page).to have_content "ordered"
+    expect(page).to have_content "Order Submitted On:"
+    expect(page).to have_content "Order Updated On:"
+  end
```
Comment (Yoyo2Code) [inline spec/good_tests/features/user_can_view_past_orders_spec.rb]: High number of validations for booking. Good job.

File: spec/good_tests/features/user_can_make_a_reservation_spec.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
+require 'rails_helper'
+
+RSpec.describe 'User can make a reservation', type: :feature do
+  scenario 'they click checkout in the cart and the reservation is made' do
+    user = create(:user)
+    business = create(:business)
+    location = create(:location) do |loc|
+      loc.properties.create(attributes_for(:property, business_id: business.id))
+    end
```
Comment (pwentz) [inline spec/good_tests/features/user_can_make_a_reservation_spec.rb]: I've never seen this syntax, before. I like it. Another way you could write it is
```
create(:property, business: business, location: location)
```

File: spec/good_tests/features/user_can_view_past_orders_spec.rb (hunk 8 above)
Comment (pwentz) [inline spec/good_tests/features/user_can_view_past_orders_spec.rb]: Are all those required for the order to be created?

**Comment by dtinianow** [comment]
Great feedback, thanks all

**Pull request merged**